Factor out ChannelBuffer from BytesReference

s1monw opened the pull request:
The ChannelBuffer interface today leaks into the BytesReference abstraction
which causes a hard dependency on netty across the board. This chance moves
this dependency and all BytesReference -> ChannelBuffer conversion into
NettyUtlis and removes the abstraction leak on BytesReference.
This change also removes unused methods on the BytesReference interface
and simplifies access to internal pages.

s1monw commented:
@jasontedor, @rmuir or @rjernst can anybody take a look at this

rjernst commented on core/src/main/java/org/elasticsearch/common/bytes/PagedBytesReference.java:
```diff
@@ -451,4 +358,30 @@ public int available() throws IOException {
+            public BytesRef next() throws IOException {
+                if (position == 0 && offset == 0 && length > 0) {
+                    // we are a slice and we are starting somewhere in the middle of a page
+                    int fragmentSize = Math.min(length, PAGE_SIZE - (offset % PAGE_SIZE));
```
isn't offset always 0 here, based on the `if` condition? So this substraction is always 0?

rjernst replied:
I think it should be `offset != 0`?

s1monw replied:
you are right that is wrong. The problem is that if we don't do the page alignment we might get copies of the arrays which we want to prevent. I `byteArray.get` returns a boolean if it materialized the array I will add assertions that ensures we don't materialize! good catch

rjernst commented:
This looks ok to me. The one comment I left I think indicates the special case for slice in the iterator is not necessary? The randomized test looks correct to me. I think without the special case, it just means it is still going by page, but it is not the same pages of bytes you would get if you started without an offset (to me that is ok, I don't think page alignment matters there, but maybe that was the intent and I am wrong).

s1monw commented:
@rjernst I simplified the iteration and added assertions. The good news is that the test failed immediately after I added the assertions. I will push this in a bit

jpountz commented on core/src/test/java/org/elasticsearch/common/netty/NettyUtilsTests.java:
```diff
@@ -0,0 +1,107 @@
+                out.writeByte((byte) random().nextInt(1 << 8));
+            }
+        } catch (IOException e) {
+            fail("should not happen " + e.getMessage());
```
should we just let the exception propagate?

s1monw replied:
copy paste... I will do that

jpountz commented:
LGTM!
